models/db/engine.go

```diff
@@ -293,3 +293,25 @@ func GetMaxID(beanOrTableName interface{}) (maxID int64, err error) {
 	_, err = x.Select("MAX(id)").Table(beanOrTableName).Get(&maxID)
 	return
 }
+
+// QuoteString quotes the string for SQLs. In most cases, SQL should use the ORM builder.
+// This function could only be used in rare cases when there is no other choices.
+// It panics if the database is not supported to avoid corrupted data
+func QuoteString(ctx context.Context, s string) string {
+	e, ok := GetEngine(ctx).(*xorm.Engine)
+	if !ok {
+		panic("can not get the real database engine")
+	}
+
+	switch e.Dialect().URI().DBType {
+	case schemas.MYSQL:
+		s = strings.ReplaceAll(s, "\\", "\\\\")
+		s = strings.ReplaceAll(s, "'", "''")
+		return "'" + s + "'"
+	case schemas.SQLITE, schemas.POSTGRES, schemas.MSSQL:
+		s = strings.ReplaceAll(s, "'", "''")
```

Comment:
this is not enough for PostgreSQL. This would be correct way for it:
```go
		s = strings.ReplaceAll(s, "'", "''")
		if strings.Contains(s, "\\") {
			s = strings.ReplaceAll(s, "\\", "\\\\")
			s = " E'" + s + "'"
		} else {
			s = "'" + s + "'"
		}
		return s
```

Reply:
I don't see the necessary. The unit test has covered your case.


And that just matches XORM's [convertStringSingleQuote](https://gitea.com/xorm/xorm/src/commit/26d291bbc311274388045768386f015a06b3be10/internal/statements/statement_args.go#L56)

Reply:
it is if server has set `standard_conforming_strings` configuration to `off` than for example: `select 'my\123test'` will actually result into `myStest` or even in some combinations in SQL syntax error. Fro example:
```
ERROR:  invalid byte sequence for encoding "UTF8": 0x8b
SQL state: 22021
```

`E''my\\123test'` will work correctly independently of PostgreSQL configuration.

All your test cases are run on single DB version with single configuration so they are just covering in best case that db servers versions default behavior.

Also from security best practices perspective using manual quoting is not something to do, that's why there is variable binding. If we really want to add it must be done in secure way

Reply:
That's user's problem. Just like MySQL's NO_BACKSLASH_ESCAPES, PgSQL has its own quirk [backslash_quote and standard_conforming_strings](https://www.postgresql.org/docs/current/runtime-config-compatible.html). 

At the moment, XORM's `convertStringSingleQuote` and `convertString` are quite simple. If you want to make everything perfectly correct, then XORM should do the fix and expose a correct string quoting function which respects various database quirks.

`db.QuoteString` shouldn't pay the cost for that.

Reply:
If you really worry about end users doing something stupid with their databases, Gitea can run some SQL checks during startup to see if the database quirks are acceptable, to prevent from incompatible database configs (that's another topic).

Reply:
Xorm will actually mostly use variable binding and this way it will work without problems as the driver will actually do string quote!

Reply:
Have you read the XORM code about `convertStringSingleQuote` and `convertString`?

Reply:
So you proposing to add checks for supported db config just for this function to be simpler?
Third choice would be not to introduce such function at all and use variable binding as it's needs to be done in the first place. I'm quite sure that at some point this function will be abused in the future code if added

Reply:
Yes I did read xorm code and it does not uses these functions at all because of: 
https://gitea.com/xorm/xorm/src/branch/master/internal/statements/statement_args.go#L78

It will add placeholder and use argument variable binding thus database driver library will do string escaping

Reply:
> So you proposing to add checks for supported db config just for this function to be simpler?
> 
> Third choice would be not to introduce such function at all and use variable binding as it's needs to be done in the first place. I'm quite sure that at some point this function will be abused in the future code if added.
> 
> Update: this is the reason why drivers/frameworks do not provide this function as public as this is not secure and can lead not only to errors but also security vulnerabilities

For other choice like `QuoteStringStrict`, #19816
* https://github.com/go-gitea/gitea/pull/19816

That's what I can do now. I am open, feel free to propose your PRs to replace these two PRs.

Reply:
> Update: this is the reason why drivers/frameworks do not provide this function as public as this is not secure and can lead not only to errors but also security vulnerabilities

Most frameworks do provide, like Python, PHP, etc. Only Golang driver is conservative and dogmatic. I do not see why exposing a correct quoting function will cause security problems. If the driver/framework doesn't, users just want to write their own unsafe quoting functions, that's a more serious problem.